finality-aleph/src/metrics/finality_rate.rs

```diff
@@ -0,0 +1,162 @@
+use std::collections::{hash_map::Entry, HashMap};
+
+use primitives::{BlockHash, BlockNumber};
+use substrate_prometheus_endpoint::{register, Counter, PrometheusError, Registry, U64};
+
+#[derive(Clone)]
+pub enum FinalityRateMetrics {
+    Prometheus {
+        own_finalized: Counter<U64>,
+        own_hopeless: Counter<U64>,
+        imported_cache: HashMap<BlockNumber, Vec<BlockHash>>,
+    },
+    Noop,
+}
+
+impl FinalityRateMetrics {
+    pub fn new(registry: Option<&Registry>) -> Result<Self, PrometheusError> {
+        let registry = match registry {
+            None => return Ok(FinalityRateMetrics::Noop),
+            Some(registry) => registry,
+        };
+
+        Ok(FinalityRateMetrics::Prometheus {
+            own_finalized: register(
+                Counter::new("aleph_own_finalized_blocks", "no help")?,
+                registry,
+            )?,
+            own_hopeless: register(
+                Counter::new("aleph_own_hopeless_blocks", "no help")?,
+                registry,
+            )?,
+            imported_cache: HashMap::new(),
+        })
+    }
+
+    /// Stores the imported block's hash. Assumes that the imported block is own.
+    pub fn report_own_imported(&mut self, hash: BlockHash, number: BlockNumber) {
+        let imported_cache = match self {
+            FinalityRateMetrics::Prometheus { imported_cache, .. } => imported_cache,
+            FinalityRateMetrics::Noop => return,
+        };
+
+        match imported_cache.entry(number) {
+            Entry::Occupied(mut entry) => entry.get_mut().push(hash),
+            Entry::Vacant(entry) => {
+                entry.insert(vec![hash]);
+            }
+        }
+    }
+
+    /// Counts the blocks at the level of `number` different than the passed block
+    /// and reports them as hopeless. If `hash` is a hash of own block it will be found
+    /// in `imported_cache` and reported as finalized.
+    pub fn report_finalized(&mut self, hash: BlockHash, number: BlockNumber) {
+        let (own_finalized, own_hopeless, imported_cache) = match self {
+            FinalityRateMetrics::Prometheus {
+                own_finalized,
+                own_hopeless,
+                imported_cache,
+            } => (own_finalized, own_hopeless, imported_cache),
+            FinalityRateMetrics::Noop => return,
+        };
+
+        match imported_cache.entry(number) {
+            Entry::Occupied(entry) => {
+                own_hopeless.inc_by(
+                    entry
+                        .get()
+                        .iter()
+                        .filter(|h| {
+                            if **h == hash {
+                                own_finalized.inc();
+                            }
+                            **h != hash
+                        })
+                        .count() as u64,
+                );
```

Comment:
This could be more readable if done in this way:
```
let hashes = entry.get();
let new_hopeless_count = iter.filter(|h| **h != hash).count();
own_hopeless.inc_by(new_hopeless_count);
own_finalized.inc_by(hashes.len() - new_hopeless_count);
```